### ISSUE TITLE
What is the best way to retry a promise?

### ISSUE THREAD

**Issue by hpbl** [google/promises#45]
On the readme is says that using promises makes it easier to retry asynchronous operations. I couldn't
find a retry method or anything of the sort, so what would be the simplest way to retry an operation using promises?

Thanks.

**Comment by shoumikhin** [comment]
Hi Hilton,

I guess the docs meant [`recover`](https://github.com/google/promises/blob/master/g3doc/index.md#recover). Do you need something else? Just provide a usage example and we're happy to consider adding a new convenience extension if needed.

**Comment by hpbl** [comment]
Hey @shoumikhin thanks for the quick response, but I don't think `recover` does what I need. My use case would be the following:

```swift
makeRequest()
    .then {  // process data and carry on }
    .catch { error in
        // Retry the request n times before giving up
    }
```

I'd like to be able to easily retry the request n times before accepting that it failed and carrying on with the error handling.

**Comment by rumnat** [comment]
It would be great to have such extension like in Rx world. If a promise fails, it tries to restart it as many times as you put in parameters.

**Comment by DazChong** [comment]
Thanks to out-of-the-box `await`! I'm using it for recursive promise.

        // e.g: each page page 10 items & total of 7 pages
        func getItemsCountAt(page: Int) -> Promise<Int> {
            let items = page > 7 ? 0 : 10
            return Promise(items)
        }

        // this function keep scanning until page with no more items
        func getAllItemsCount() -> Promise<Int> {
            return Promise<Int>(on: .global()) { () -> Int in
                var page = 1
                var totalItems = 0
                
                var pageItems = 0
                repeat {
                    pageItems = try await(getItemsCountAt(page: page))
                    totalItems += pageItems
                    page += 1
                } while pageItems > 0
                
                return totalItems
            }
        }
        
        // usage
        getAllItemsCount().then { print($0) }

**Comment by hpbl** [comment]
@shoumikhin It looks great for me! The default values are awesome for when you don't need that much control over the retry. Can't wait to use it!

**Comment by rumnat** [comment]
It would be better to have it like this:
myAsyncFunc(param)
.retry(3)
.then { value in
    
}.catch { error in
    
}

**Comment by shoumikhin** [comment]
Hi Anton, I'm afraid that won't be possible, because in our implementation a promise doesn't hold any work block associated with it, so it's not possible to execute it again retrospectively. We've intentionally designed promises that way to avoid extra retain cycles, so that the work block is owned solely by GCD most of the time, instead of a promise itself.

**Comment by temrich** [comment]
`retry` operator is now available in the latest release (1.2.3): https://github.com/google/promises/blob/master/g3doc/index.md#retry

Please let us know if you have any questions or feedback or run into any issues. Thank you!
